File: 6.硬币/go/coin.go
func waysToChange(n int) int {
	dp := make([]int, 1000000)
	dp[0] = 1
	coins := []int{1, 5, 10, 25}
	for _, coin := range coins {
		//按照一定次序，把硬币的排列情况进行累加，最后得到n的排列
		for i := coin; i <= n; i++ {
			dp[i] = dp[i] + dp[i-coin]
		}
	}
	return dp[n] % 1000000007
}

File: 06.硬币/cpp/coin.cpp
class Solution {
public:
    int waysToChange(int n) {
        const int N = 1000000;
        const int mod = 1000000007;
        int coins[4] = {1, 5, 10, 25};
        int dp[N+1] = {1};

        for (int i = 0; i < 4; ++i) {
            for (int j = coins[i]; j <= n; j++) {
                dp[j] = (dp[j] + dp[j-coins[i]]) % mod;
            }
        }
        return dp[n];
    }
};
